File: generated_tasks/graphkit__25ebacd18c75/tests/hidden/base_repo/src/flow.ts
import { DiGraph } from './digraph';
import { Graph } from './graph';
import { NodeId } from './types';
import { NodeNotFoundError } from './errors';

export interface FlowResult {
  maxFlow: number;
  flowOnEdge: Map<string, number>;
  minCut: { sourceSet: NodeId[]; sinkSet: NodeId[] };
}

function bfsPath(
  capacity: Map<string, Map<NodeId, number>>,
  source: NodeId,
  sink: NodeId,
  parent: Map<NodeId, NodeId | null>,
  nodes: NodeId[]
): boolean {
  const visited = new Set<NodeId>([source]);
  const queue: NodeId[] = [source];

  while (queue.length > 0) {
    const u = queue.shift()!;
    const caps = capacity.get(String(u));
    if (!caps) continue;

    for (const [v, cap] of caps) {
      if (!visited.has(v) && cap > 0) {
        visited.add(v);
        parent.set(v, u);
        if (v === sink) return true;
        queue.push(v);
      }
    }
  }
  return false;
}

export function fordFulkerson(
  g: DiGraph,
  source: NodeId,
  sink: NodeId
): FlowResult {
  if (!g.hasNode(source)) throw new NodeNotFoundError(source);
  if (!g.hasNode(sink)) throw new NodeNotFoundError(sink);

  const capacity = new Map<string, Map<NodeId, number>>();
  const flowOnEdge = new Map<string, number>();
  const nodes = g.nodeIds();

  for (const id of nodes) capacity.set(String(id), new Map());

  for (const edge of g.edges()) {
    const w = typeof edge.weight === 'number' ? edge.weight : 1;
    const src = String(edge.source);
    const tgt = edge.target;
    const fwdKey = `${edge.source}->${edge.target}`;
    const revKey = `${edge.target}->${edge.source}`;

    if (!capacity.has(src)) capacity.set(src, new Map());
    capacity.get(src)!.set(tgt, (capacity.get(src)!.get(tgt) ?? 0) + w);
    flowOnEdge.set(fwdKey, 0);
    flowOnEdge.set(revKey, 0);

    const tgtStr = String(edge.target);
    if (!capacity.has(tgtStr)) capacity.set(tgtStr, new Map());
    if (!capacity.get(tgtStr)!.has(edge.source)) {
      capacity.get(tgtStr)!.set(edge.source, 0);
    }
  }

  let maxFlow = 0;

  while (true) {
    const parent = new Map<NodeId, NodeId | null>();
    if (!bfsPath(capacity, source, sink, parent, nodes)) break;

    let pathFlow = Infinity;
    let v = sink;
    while (v !== source) {
      const u = parent.get(v)!;
      const cap = capacity.get(String(u))?.get(v) ?? 0;
      pathFlow = Math.min(pathFlow, cap);
      v = u;
    }

    v = sink;
    while (v !== source) {
      const u = parent.get(v)!;
      capacity.get(String(u))!.set(v, (capacity.get(String(u))!.get(v) ?? 0) - pathFlow);
      capacity.get(String(v))!.set(u, (capacity.get(String(v))!.get(u) ?? 0) + pathFlow);
      const fwdKey = `${u}->${v}`;
      flowOnEdge.set(fwdKey, (flowOnEdge.get(fwdKey) ?? 0) + pathFlow);
      v = u;
    }

    maxFlow += pathFlow;
  }

  const visited = new Set<NodeId>([source]);
  const queue: NodeId[] = [source];
  while (queue.length > 0) {
    const u = queue.shift()!;
    for (const [v, cap] of (capacity.get(String(u)) ?? new Map())) {
      if (!visited.has(v) && cap > 0) { visited.add(v); queue.push(v); }
    }
  }

  const sourceSet = [...visited];
  const sinkSet = nodes.filter(n => !visited.has(n));

  return { maxFlow, flowOnEdge, minCut: { sourceSet, sinkSet } };
}

export function maxBipartiteMatching(g: Graph): number {
  const { bipartite, partition } = (() => {
    const color = new Map<NodeId, 0 | 1>();
    const p0: NodeId[] = [], p1: NodeId[] = [];
    for (const start of g.nodeIds()) {
      if (color.has(start)) continue;
      color.set(start, 0); p0.push(start);
      const q = [start];
      while (q.length > 0) {
        const u = q.shift()!;
        for (const v of g.neighbors(u)) {
          if (!color.has(v)) {
            const c = (1 - color.get(u)!) as 0 | 1;
            color.set(v, c); (c === 0 ? p0 : p1).push(v); q.push(v);
          } else if (color.get(v) === color.get(u)) return { bipartite: false, partition: undefined };
        }
      }
    }
    return { bipartite: true, partition: [p0, p1] as [NodeId[], NodeId[]] };
  })();

  if (!bipartite || !partition) return 0;

  const [left, right] = partition;
  const matchL = new Map<NodeId, NodeId | null>(left.map(id => [id, null]));
  const matchR = new Map<NodeId, NodeId | null>(right.map(id => [id, null]));

  const dfs = (u: NodeId, visited: Set<NodeId>): boolean => {
    for (const v of g.neighbors(u)) {
      if (visited.has(v)) continue;
      visited.add(v);
      if (matchR.get(v) === null || matchR.get(v) === undefined || dfs(matchR.get(v)!, visited)) {
        matchL.set(u, v);
        matchR.set(v, u);
        return true;
      }
    }
    return false;
  };

  let count = 0;
  for (const u of left) {
    const visited = new Set<NodeId>();
    if (dfs(u, visited)) count++;
  }
  return count;
}

export function minCutEdges(g: DiGraph, source: NodeId, sink: NodeId): [NodeId, NodeId][] {
  const result = fordFulkerson(g, source, sink);
  const { sourceSet, sinkSet } = result.minCut;
  const sinkSet2 = new Set(sinkSet);
  const cuts: [NodeId, NodeId][] = [];
  for (const edge of g.edges()) {
    if (sourceSet.includes(edge.source) && sinkSet2.has(edge.target)) {
      cuts.push([edge.source, edge.target]);
    }
  }
  return cuts;
}

File: generated_tasks/cryograph__mincut-residual-partition/environment/repo/src/flow/flow.cpp
#include "flow/flow.hpp"
#include <algorithm>
#include <queue>
#include <limits>

namespace cryo {

FlowResult edmonds_karp(const Graph& g, NodeId source, NodeId sink) {
    FlowResult result;
    result.max_flow = 0.0;

    if (!g.has_node(source) || !g.has_node(sink) || source == sink)
        return result;

    auto ids = g.node_ids();
    size_t n = ids.size();
    std::unordered_map<NodeId, size_t> id_to_idx;
    for (size_t i = 0; i < n; i++) id_to_idx[ids[i]] = i;

    std::vector<std::vector<double>> capacity(n, std::vector<double>(n, 0.0));
    std::vector<std::vector<double>> flow(n, std::vector<double>(n, 0.0));

    g.for_each_edge([&](const Edge& e) {
        size_t u = id_to_idx[e.src], v = id_to_idx[e.dst];
        capacity[u][v] += e.weight;
    });

    size_t si = id_to_idx[source], ti = id_to_idx[sink];

    while (true) {
        std::vector<int> parent(n, -1);
        parent[si] = static_cast<int>(si);
        std::queue<size_t> q;
        q.push(si);

        while (!q.empty() && parent[ti] == -1) {
            size_t u = q.front();
            q.pop();
            for (size_t v = 0; v < n; v++) {
                if (parent[v] == -1 && capacity[u][v] - flow[u][v] > 1e-12) {
                    parent[v] = static_cast<int>(u);
                    q.push(v);
                }
            }
        }

        if (parent[ti] == -1) break;

        double path_flow = std::numeric_limits<double>::infinity();
        for (size_t v = ti; v != si; v = static_cast<size_t>(parent[v])) {
            size_t u = static_cast<size_t>(parent[v]);
            path_flow = std::min(path_flow, capacity[u][v] - flow[u][v]);
        }

        for (size_t v = ti; v != si; v = static_cast<size_t>(parent[v])) {
            size_t u = static_cast<size_t>(parent[v]);
            flow[u][v] += path_flow;
            flow[v][u] -= path_flow;
        }

        result.max_flow += path_flow;
    }

    g.for_each_edge([&](const Edge& e) {
        size_t u = id_to_idx[e.src], v = id_to_idx[e.dst];
        if (flow[u][v] > 1e-12) {
            result.edge_flow[e.id] = flow[u][v];
        }
    });

    return result;
}

MinCutResult min_cut(const Graph& g, NodeId source, NodeId sink) {
    MinCutResult result;
    if (!g.has_node(source) || !g.has_node(sink)) return result;

    auto ids = g.node_ids();
    size_t n = ids.size();
    std::unordered_map<NodeId, size_t> id_to_idx;
    for (size_t i = 0; i < n; i++) id_to_idx[ids[i]] = i;

    std::vector<std::vector<double>> capacity(n, std::vector<double>(n, 0.0));
    std::vector<std::vector<double>> flow(n, std::vector<double>(n, 0.0));

    g.for_each_edge([&](const Edge& e) {
        capacity[id_to_idx[e.src]][id_to_idx[e.dst]] += e.weight;
    });

    size_t si = id_to_idx[source], ti = id_to_idx[sink];
    result.cut_value = 0.0;

    while (true) {
        std::vector<int> parent(n, -1);
        parent[si] = static_cast<int>(si);
        std::queue<size_t> q;
        q.push(si);
        while (!q.empty() && parent[ti] == -1) {
            size_t u = q.front(); q.pop();
            for (size_t v = 0; v < n; v++) {
                if (parent[v] == -1 && capacity[u][v] - flow[u][v] > 1e-12) {
                    parent[v] = static_cast<int>(u);
                    q.push(v);
                }
            }
        }
        if (parent[ti] == -1) break;
        double pf = std::numeric_limits<double>::infinity();
        for (size_t v = ti; v != si; v = (size_t)parent[v])
            pf = std::min(pf, capacity[(size_t)parent[v]][v] - flow[(size_t)parent[v]][v]);
        for (size_t v = ti; v != si; v = (size_t)parent[v]) {
            flow[(size_t)parent[v]][v] += pf;
            flow[v][(size_t)parent[v]] -= pf;
        }
        result.cut_value += pf;
    }

    std::vector<bool> visited(n, false);
    std::queue<size_t> bfs;
    bfs.push(si);
    visited[si] = true;
    while (!bfs.empty()) {
        size_t u = bfs.front(); bfs.pop();
        result.source_side.insert(ids[u]);
        for (size_t v = 0; v < n; v++) {
            if (!visited[v] && capacity[u][v] > 1e-12) {
                visited[v] = true;
                bfs.push(v);
            }
        }
    }
    for (size_t i = 0; i < n; i++) {
        if (!visited[i]) result.sink_side.insert(ids[i]);
    }

    g.for_each_edge([&](const Edge& e) {
        if (result.source_side.count(e.src) && result.sink_side.count(e.dst)) {
            result.cut_edges.push_back(e);
        }
    });

    return result;
}

double max_flow_value(const Graph& g, NodeId source, NodeId sink) {
    return edmonds_karp(g, source, sink).max_flow;
}

} // namespace cryo